File: packages/sdui/src/main/java/nl/q42/sdui/ScreenRegistry.java
package nl.q42.sdui;

import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

import nl.q42.common.screen.Screen;
import nl.q42.common.screen.ScreenTab;
import nl.q42.sdui.screen.HomeScreen;
import nl.q42.sdui.screen.ProfileScreen;
import nl.q42.sdui.screen.SearchScreen;
import nl.q42.sdui.screen.SettingsScreen;

@Service

public class ScreenRegistry {
    private final List<Screen> screens;

    public static final ScreenTab[] SCREEN_TABS = new ScreenTab[]{
            new ScreenTab("tab.home", "home.png", HomeScreen.SCREEN_NAME),
            new ScreenTab("tab.search", "search.png", SearchScreen.SCREEN_NAME),
            new ScreenTab("tab.profile", "profile.png", ProfileScreen.SCREEN_NAME),
            new ScreenTab("tab.settings", "settings.png", SettingsScreen.SCREEN_NAME)
    };

    public ScreenRegistry(List<Screen> screens) {
        this.screens = screens;
    }

    public List<Screen> getAll() {
        return this.screens;
    }

    public Screen getByIdentifier(String screenIdentifier) {
        return this.screens.stream()
                .filter(screen -> screen.name.equals(screenIdentifier))
                .findFirst()
                .orElse(null);
    }
}

File: packages/sdui/src/main/kotlin/nl/q42/sdui/ScreenRegistry.kt
package nl.q42.sdui

import nl.q42.common.screen.Screen
import nl.q42.common.screen.ScreenTab
import nl.q42.core.AppRequestContext
import nl.q42.sdui.screen.*
import org.springframework.stereotype.Service

@Service
class ScreenRegistry(private val screens: MutableList<ScreenInstance>) {

    val all: MutableList<ScreenInstance>
        get() = this.screens

    fun getByIdentifier(screenIdentifier: String, context: AppRequestContext): Screen? {
        return this.screens.stream()
            .filter { screen: ScreenInstance? -> screen?.name().equals(screenIdentifier) }
            .findFirst()
            .map { screen: ScreenInstance -> screen.createScreen(context) }
            .orElse(null)
    }

    fun defaultScreen(context: AppRequestContext): Screen {
        val initial = getByIdentifier(DEFAULT_SCREEN_IDENTIFIER, context)

        return when {
            initial != null -> initial
            screens.isNotEmpty() -> screens[0].createScreen(context)
            else -> throw IllegalStateException("No screens available in the registry")
        }

    }

    companion object {
        val SCREEN_TABS: List<ScreenTab> = listOf<ScreenTab>(
            ScreenTab("Home", "home.png", HomeScreen.SCREEN_IDENTIFIER),
            ScreenTab("Search", "search.png", SearchScreen.SCREEN_IDENTIFIER),
            ScreenTab("Profile", "profile.png", ProfileScreen.SCREEN_IDENTIFIER),
            ScreenTab("Settings", "settings.png", SettingsScreen.SCREEN_IDENTIFIER)
        )

        const val DEFAULT_SCREEN_IDENTIFIER = HomeScreen.SCREEN_IDENTIFIER;
    }
}
